internal/coord/routing/crawl.go

```diff
@@ -0,0 +1,357 @@
+			cpls:    map[string]int{},
+			waiting: map[string]N{},
+			success: map[string]N{},
+			failed:  map[string]N{},
+			errors:  map[string]error{},
+		}
+
+		for _, node := range tev.Seed {
+			// exclude self from closest nodes
+			if key.Equal(node.Key(), c.self.Key()) {
+				continue
+			}
+
+			for j := 0; j < c.cfg.MaxCPL; j++ {
+				target, err := c.cplFn(node.Key(), j)
+				if err != nil {
+					// return nil, fmt.Errorf("generate cpl: %w", err)
+					// TODO: log
+					continue
+				}
+
+				job := crawlJob[K, N]{
+					node:   node,
+					target: target.Key(),
+				}
+
+				ci.cpls[job.mapKey()] = j
+				ci.todo = append(ci.todo, job)
+			}
+		}
+
+		c.info = ci
+
+	case *EventCrawlNodeResponse[K, N]:
+		span.SetAttributes(attribute.Int("closer_nodes", len(tev.CloserNodes)))
+
+		if c.info == nil {
+			return &StateCrawlIdle{}
+		}
+
+		job := crawlJob[K, N]{
+			node:   tev.NodeID,
+			target: tev.Target,
+		}
+
+		mapKey := job.mapKey()
+		if _, found := c.info.waiting[mapKey]; !found {
+			break
+		}
+
+		delete(c.info.waiting, mapKey)
+		c.info.success[mapKey] = tev.NodeID
+
+		for _, node := range tev.CloserNodes {
+			for i := 0; i < c.cfg.MaxCPL; i++ {
+				target, err := c.cplFn(node.Key(), i)
+				if err != nil {
+					// TODO: log
+					continue
+				}
+
+				newJob := crawlJob[K, N]{
+					node:   node,
+					target: target.Key(),
+				}
+
+				newMapKey := newJob.mapKey()
+				if _, found := c.info.cpls[newMapKey]; found {
+					continue
+				}
+
+				c.info.cpls[newMapKey] = i
+				c.info.todo = append(c.info.todo, newJob)
+			}
+		}
+
+	case *EventCrawlNodeFailure[K, N]:
+		if c.info == nil {
+			return &StateCrawlIdle{}
+		}
+
+		job := crawlJob[K, N]{
+			node:   tev.NodeID,
+			target: tev.Target,
+		}
+
+		mapKey := job.mapKey()
+		if _, found := c.info.waiting[mapKey]; !found {
+			break
+		}
+		span.RecordError(tev.Error)
+
+		delete(c.info.waiting, mapKey)
+		c.info.failed[mapKey] = tev.NodeID
+		c.info.errors[mapKey] = tev.Error
+
+	case *EventCrawlPoll:
+		// no event to process
+	default:
+		panic(fmt.Sprintf("unexpected event: %T", tev))
+	}
+
+	if c.info == nil {
+		return &StateCrawlIdle{}
+	}
+
+	if len(c.info.waiting) >= c.cfg.MaxCPL*c.cfg.Concurrency {
+		return &StateCrawlWaitingAtCapacity{}
+	}
+
+	if len(c.info.todo) > 0 {
+
+		// pop next crawl job from queue
+		var job crawlJob[K, N]
+		job, c.info.todo = c.info.todo[0], c.info.todo[1:]
+
+		// mark the job as waiting
+		mapKey := job.mapKey()
+		c.info.waiting[mapKey] = job.node
+
+		return &StateCrawlFindCloser[K, N]{
+			Target: job.target,
+			NodeID: job.node,
+		}
+	}
+
+	if len(c.info.waiting) > 0 {
+		return &StateCrawlWaitingWithCapacity{}
+	}
+
+	// generate list of new nodes for the routing table
+	nodes := make([]N, len(c.info.success))
+	i := 0
+	for _, node := range c.info.success {
+		nodes[i] = node
+		i += 1
+	}
+
+	// clear info to indicate that we're idle
+	c.info = nil
+
+	return &StateCrawlFinished[K, N]{
+		Nodes: nodes,
```

Comment:
This could contain thousands of nodes, do we need to return this? Could include stats instead: number found, number of errors etc.